python/mxnet/kvstore.py

```diff
@@ -349,6 +349,77 @@ def row_sparse_pull(self, key, out=None, priority=0, row_ids=None):
+        can reduce communication costs. The difference between these thresholded values and
+        original values is stored at the sender's end as residual and added to the
+        gradient in the next iteration.
+
+        When kvstore is 'local', gradient compression is used to reduce communication
+        between multiple devices (gpus). Gradient is quantized on each GPU which
+        computed the gradients, then sent to the GPU which merges the gradients. This
+        receiving GPU dequantizes the gradients and merges them. Note that this
+        increases memory usage on each GPU because of the residual array stored.
+
+        When kvstore is 'dist', gradient compression is used to reduce communication
+        from worker to sender. Gradient is quantized on each worker which
+        computed the gradients, then sent to the server which dequantizes
+        this data and merges the gradients from each worker. Note that this
+        increases CPU memory usage on each worker because of the residual array stored.
+        Only worker to server communication is compressed in this setting.
+        If each machine has multiple GPUs, currently this GPU to GPU communication is
+        not compressed. Server to worker communication (in the case of pull) is also not
+        compressed.
+
+        To use 2bit compression, we need to specify `compression` as `2bit`.
+        Only specifying `compression` would use default value for the threshold.
+        To completely specify the arguments for 2bit compression, we would need to pass
+        a dictionary which includes `threshold` like:
+        {'compression': '2bit', 'threshold': 0.5}
+
+        Parameters
+        ----------
+        compression_params : dict
+            A dictionary specifying the type and parameters for gradient compression.
+            The key `compression` in this dictionary is a
+            required string argument and specifies the type of gradient compression.
+            Other keys in this dictionary are optional and specific to the type
+            of gradient compression. Defaults to (('compression', '2bit'),).
+            The default value is not a dict,
+            just to avoid pylint warning on dangerous default values.
+        """
+        if compression_params:
+            if not isinstance(compression_params, dict):
+                raise ValueError("compression_params needs to be a dictionary")
+            if 'compression' not in compression_params:
+                raise ValueError('compression_params requires `compression` to be set')
+            elif not isinstance(compression_params['compression'], string_types):
+                raise TypeError('compression must be a string')
+            elif compression_params['compression'] not in ['none', '2bit']:
+                raise ValueError('Unsupported type of compression')
+
+            if compression_params['compression'] == '2bit':
```

Comment:
These parsing should be done in backend with `dmlc::Parameter`

The frontend should pass strings of key value pairs.